Add Japanese layer

kenjimyzk opened the pull request:
Similar to Chinese Layer, I add Japanese Layer as a new Layer.
Please check that.
Thank you.

d12frosted commented on layers/+intl/japanese/packages.el:
```diff
@@ -0,0 +1,93 @@
+;; List of all packages to install and/or initialize. Built-in packages
+;; which require an initialization must be listed explicitly in the list.
+(setq japanese-packages
+      '(helm
```
Here you state that your layer uses `helm`, but there is no `japanese/[post|pre]-?init-helm` function. So this is redundant.

d12frosted commented on layers/+intl/japanese/packages.el:
```diff
@@ -0,0 +1,93 @@
+    (setq migemo-coding-system 'utf-8-unix)
+    (migemo-init)
+    (with-eval-after-load "helm"
+      (helm-migemo-mode 1))
```
AFAIK, `helm-migemo-mode` comes from `helm-migemo` package, not the `migemo`. So you have to add `helm-migemo` to `japanese-packages` and configure it via `japanese/init-helm-migemo` function.

d12frosted replied:
Also note, that `helm-migemo` should be toggled on only when `helm` layer is used.

d12frosted commented on layers/+intl/japanese/packages.el:
```diff
@@ -0,0 +1,93 @@
+    (add-hook 'calendar-today-visible-hook 'japanese-holiday-mark-weekend)
+    (add-hook 'calendar-today-invisible-hook 'japanese-holiday-mark-weekend)
+    (add-hook 'calendar-today-visible-hook 'calendar-mark-today)
+    )
```
Please, don't leave closing parenthesis alone 😸 Move them to previous line.

d12frosted commented:
Overall, don't forget to use `:defer t` or anything else that defers package loading where it makes sense.

d12frosted commented on layers/+intl/japanese/README.org:
```diff
@@ -0,0 +1,74 @@
+file.
+
+** Configuration
+For a basic configuration, add this to your =~/.spacemacs= inside ~dotspacemacs/user-config~:
```
Please read [conventions on spacing in documentation](https://github.com/syl20bnr/spacemacs/blob/develop/doc/CONVENTIONS.org#spacing-in-documentation). Not critical, but anyway.

d12frosted commented on layers/+intl/japanese/README.org:
```diff
@@ -0,0 +1,74 @@
+     - [[pangu-spacing][pangu-spacing]]
+ - [[Key Bindings][Key Bindings]]
+
+* Description
```
Also, don't forget to fill paragraphs.

d12frosted commented:
Once you've done with everything - please squash all commits into one.

d12frosted commented on layers/+intl/japanese/packages.el:
```diff
@@ -0,0 +1,93 @@
+;;
+;; Copyright (c) 2012-2016 Sylvain Benner & Contributors
+;;
+;; Author: Sylvain Benner <sylvain.benner@gmail.com>
```
It's [OK to change](https://github.com/syl20bnr/spacemacs/blob/develop/CONTRIBUTING.org#author-of-a-new-layer) author name to yours name 😸

JAremko commented:
@kenjimyzk  thanks a lot for pitching in :heart:  

And thanks @d12frosted  for guiding.

kenjimyzk commented:
Thank @d12frosted  for everything. I will fix them asap.

kenjimyzk commented:
@d12frosted, I have two replies:
>AFAIK, helm-migemo-mode comes from helm-migemo package, not the migemo. So you have to add helm-migemo to japanese-packages and configure it via japanese/init-helm-migemo function. 
1. helm itself has support with mimeo, so helm-migemo.el is no longer required. https://github.com/emacs-helm/helm/wiki/Migemo

> Overall, don't forget to use :defer t or anything else that defers package loading where it makes sense. 
2. ~;defer t~. can be added to migemo and ddskk.  However, others do not work if ~:defer t~ is added.

kenjimyzk commented:
I have updated Japanese Layer. As well as correcting typos and errors, I added a new package, [evil-tutor-ja](https://github.com/kenjimyzk/evil-tutor-ja), which is Japanese vimtutor adapted for Evil and wrapped in a major mode.

d12frosted commented on layers/+intl/japanese/config.el:
```diff
@@ -0,0 +1,13 @@
+
+;; Variables
+
+(defvar migemo-dictionary "/usr/local/share/migemo/utf-8/migemo-dict")
```
`migemo-dictionary` is defined in [migemo](https://github.com/emacs-jp/migemo/blob/master/migemo.el#L80). So you don't have to redefine it here. 

Also it's a good thing to document variables defined in `config`.

kenjimyzk replied:
I deleted config.el.

d12frosted commented on layers/+intl/japanese/packages.el:
```diff
@@ -0,0 +1,94 @@
+    (migemo-init)
+    (japanese/post-init-helm-mode)))
+
+(defun japanese/post-init-helm-mode ()
```
Please, don't use such names for functions that are not supposed to initialise packages listed in `japanese-packages`. Since that function is called once - you can just inline it where it's called. Otherwise - please define all helper functions in `funcs.el` file.

d12frosted replied:
And again, `helm-migemo` is [separate package](https://melpa.org/#/helm-migemo) which is not downloaded with `migemo`. Please add `helm-migemo` to `japanese-packages` and setup just as you did with `avy-migemo`.

kenjimyzk replied:
[Here](https://github.com/emacs-helm/helm/wiki/Migemo) it is said:

> You will find in a package called helm-migemo.el in MELPA. This package is deprecated and not supported by helm, so please do not use it!

d12frosted commented on layers/+intl/japanese/packages.el:
```diff
@@ -0,0 +1,94 @@
+(defun japanese/init-japanese-holidays ()
+  (use-package japanese-holidays
+    :config
+            (setq calendar-holidays
```
Something is wrong with indentation here.

kenjimyzk replied:
I have fixed it.

d12frosted commented on layers/+intl/japanese/packages.el:
```diff
@@ -0,0 +1,94 @@
+(defun japanese/init-avy-migemo ()
+  (use-package avy-migemo
+    :config
+      (avy-migemo-mode 1)))
```
Something is wrong with indentation here.

d12frosted commented on layers/+intl/japanese/packages.el:
```diff
@@ -0,0 +1,94 @@
+(defun japanese/init-pangu-spacing ()
+  (use-package pangu-spacing
+    :init (progn ;; replacing `chinese-two-byte' by `japanese'
+                 (setq pangu-spacing-chinese-before-english-regexp
```
Something is wrong with indentation here.

kenjimyzk replied:
I have fixed it.

d12frosted commented on layers/+intl/japanese/README.org:
```diff
@@ -0,0 +1,79 @@
+#+TITLE: Japanese layer
```
Please checkout [conventions on spacing in documentation](https://github.com/syl20bnr/spacemacs/blob/develop/doc/CONVENTIONS.org#spacing-in-documentation) and fix any spacing problems.

kenjimyzk replied:
Thanks everything, d12frosted!!
I have updated Japanese Layer again.
Please check them.

kenjimyzk commented:
Thanks everything, d12frosted!! I have updated Japanese Layer again.
Please check them.

dudelson commented:
@d12frosted is this PR ready? Very excited about this layer finally landing in spacemacs. Let me know if there's anything I can do to help speed it along.

d12frosted commented:
I think it is, but @syl20bnr should approve before it can be merged. Sorry 😸

niku commented:
Do you have any update on this PR?

syl20bnr commented:
Thank you ! 👍 
Cherry-picked into develop branch, you can safely delete your branch.
